ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToNode.hs

```diff
@@ -510,10 +525,26 @@ mkApps kernel Tracers {..} Codecs {..} genChainSyncTimeout Handlers {..} =
       -> remotePeer
       -> Channel m bKA
       -> m ((), Maybe bKA)
-    aKeepAliveClient _version _them _channel = do
+    aKeepAliveClient version them channel = do
       labelThisThread "KeepAliveClient"
-      -- TODO implement the client
-      forever (threadDelay 1000) >> return ((), Nothing)
+      startTs <- newTVarM Nothing
+      let kacApp = case version of
+                        -- Version 1 doesn't support keep alive protocol but Blockfetch
+                        -- still requires a PeerGSV per peer.
+                        NodeToNodeV_1 -> \_ -> forever (threadDelay 1000) >> return ((), Nothing)
+                        NodeToNodeV_2 -> \_ -> forever (threadDelay 1000) >> return ((), Nothing)
+                        _             -> \dqCtx -> do
+                          runPeerWithLimits
+                            nullTracer
+                            cKeepAliveCodec
+                            (byteLimitsKeepAlive (const 0)) -- TODO: Real Bytelimits, see #1727
```

Comment:
I closed #1727 recently, did I jumped ahead?

Reply:
Yes, #1727 has not been implemented and the "TODO" is present on all NodeToNode protocols.

Reply:
I re-opened it, I think I confused it with per-state time limits.